### PR TITLE
Added back md5 hash of just code and full driver, added DCB length

### PR DESCRIPTION
**Pull request by paulhagstrom**
Looked into structure again, since everything is relative, the link field is not the problem I thought it was. That is to say, it is not going to be sensitive to which order the drivers appear within the SOS.DRIVER file. This now creates two md5 hashes, one over the bytes from the entry point to the end of the driver (which should be stable across any two drivers that are in fact the same), and one over the whole driver area which includes the comment string, slot, active status, parameter block. This way we should be able to identify when two drivers are the same but yet have different configurations. .RS232 is in fact correctly marked as different between WAP TEL 06b and TEL 10b SOS.DRIVERs, because although their code shares the same md5, there is one byte that differs within the device configuration block, so the overall driver md5s do not match. They were set up slightly differently, and that information could also be useful to distinguish.

I also added DCB length as an additional field, since that wasn't being captured before, and rather than throw away the link to the next driver, I store it in the driver structure so it can be used to set the boundaries of the "code" portion (from the entry point to the start of the next driver).

### CODE REVIEW

**Comment by paulhagstrom** [comment]
Oh, and, I also kind of "unfactored" the hash because I need to pull two different regions of bytes, it didn't seem like it was clearer to have it split out. It seems sensible to include the code block as part of the overall driver hash (rather than having a hash for the metadata and config data only, without including the code), so that means the code portion needs to go into both hashes.  It could be done more modularly, by reading in the whole driver and hashing it, then going back to entry point and re-reading the bytes from there to the end of the driver and hashing them. I don't have much of an opinion either way. I took out the separate hashing function earlier when I thought I needed to skip over the link reference and never tried to put it back after I realized that it wasn't necessary to skip over the link.

**Comment by thecompu** [comment]
At first glance, I really like what you've done here and appreciate it. I want to take a bit of time to understand it before I commit. I hope that's okay.

**Comment by paulhagstrom** [comment]
Oh sure, no problem. Let me know if there's anything in there you're wondering about.

File: Driv3rs.py [hunk 1]
```diff
@@ -246,14 +239,25 @@ def md5hash(bytes):
...
 
-    # calculate distance from beginning of current driver to
-    # next driver. reposition pointer to beginning of current
-    # driver. send number of bytes to read to hashing function
-    ## DISABLED FOR DEBUGGING ## EXPECT BLANK FIELDS IN the CSV
-    #bytes_to_read = drivers_list[i]['comment_start'] + \
-    #drivers_list[+1]['comment_start'] - 4
-    #SOSfile.seek(drivers_list[i]['comment_start'] + 4,0)
-    #drivers_list[i]['md5'] = md5hash(bytes_to_read)
+    # device configuration block length
```
Comment (thecompu) [inline Driv3rs.py:242]: According to the official manual, the Configuration Block is optional. Also, I don't see anything that indicates a common byte-length, so I'm not clear on where you're getting 2 bytes from...? -- Never mind, I see what you're getting at. I'll write more in the main thread.

**Comment by thecompu** [comment]
Can you explain a little more about the DCB addition?
I take it that you feel it's worth adding that to the md5 even though the user can edit it and it's also optional. This kind of makes me wonder about the general philosophy of md5 hash generation.

Should the md5 hash represent:
1. Only items that would be hard-coded by the manufacturer? (e.g. mfg. byte, code block,  etc.)
2. Items that are hard-coded and some items that are user-edited? (e.g. the above plus comment (which can be edited), the dcb (which can often be edited, if it exists), etc.)

Part of me thinks that it's most useful to include only those items users can't change. Meaning, a duplicate md5 represents a manufacturer-written driver whose parts that contiribute to the hash could not have been edited through the SCP. They are technically the same as the manufacturer intended.

What is your feeling? I'm not married to either view.

File: Driv3rs.py (hunk 1 above)
Comment (paulhagstrom) [inline Driv3rs.py:242]: Yep, page 36. The first two bytes of the DIB Configuration Block contain the number of bytes in the block.

**Comment by paulhagstrom** [comment]
My thought on this is a bit more toward enabling archaeology. So, strictly speaking, we could get by with just the "code md5" that contains nothing specifically user-editable (though it does exclude some things that are not user-editable, like the manufacturer ID). However, if we have a driver, no documentation, and find three different instances with different configurations, it could be useful to try to "reverse engineer" the configuration parameters by looking at the different kinds of configurations that exist. So, the "driver md5" is sensitive even to user-editable changes in configuration so as not to collapse those distinctions. That was the main reason I wanted to keep a full-driver hash. (This also applies not just to the DCB but also to the comment, which definitely has useful information, but is directly editable in the SCP.)

**Comment by thecompu** [comment]
Thanks! I like this version a bit more for a lot of reasons.

**Pull request merged**